File: leetcode/array/trapping_rain_water.py
# https://leetcode.com/problems/trapping-rain-water/

# time complexity: O(n)

class Solution:
    def trap(self, height: list[int]) -> int:
        if len(height) < 3:
            return 0
        
        left_max = height[(left := 0)]
        right_max = height[(right := len(height) - 1)]
        volume = 0
        while left < right:
            if left_max < right_max:
                volume += left_max - height[left]
                left += 1
                left_max = max(left_max, height[left])
            else:
                volume += right_max - height[right]
                right -= 1
                right_max = max(right_max, height[right])
        return volume

File: leetcode/array/trapping_rain_water.cc
// https://leetcode.com/problems/trapping-rain-water/

// time complexity: O(n)

#include <vector>

using namespace std;

class Solution {
public:
    int trap(vector<int> &height) {
        int len = height.size();
        if (len < 3)
            return 0;
        int volume = 0, left = 0, right = len - 1;
        int left_max = height[left], right_max = height[right];
        while (left < right) {
            if (left_max < right_max) {
                volume += left_max - height[left++];
                left_max = max(left_max, height[left]);
            } else {
                volume += right_max - height[right--];
                right_max = max(right_max, height[right]);
            }
        }
        return volume;
    }
};